libafl_libfuzzer/libafl_libfuzzer_runtime/src/fuzz.rs

```diff
@@ -135,33 +168,80 @@ fn fuzz_single(
+    let std_power = SkippableStage::new(std_power, |_| mutator_status.std_mutational.into());
+
+    // for custom mutator and crossover, each have access to the LLVMFuzzerMutate -- but it appears
+    // that this method doesn't normally offer stacked mutations where one may expect them
+    // we offer stacked mutations since this appears to be expected; see:
+    // https://github.com/google/fuzzing/blob/bb05211c12328cb16327bb0d58c0c67a9a44576f/docs/structure-aware-fuzzing.md#example-compression
+    // additionally, we perform mutation and crossover in two separate stages due to possible
+    // errors introduced by incorrectly handling custom mutations; see explanation below
+
+    // a custom mutator is defined
+    // note: in libfuzzer, crossover is enabled by default, but this appears to be unintended
+    // and erroneous if custom mutators are defined as it inserts bytes from other test cases
+    // without performing the custom mutator's preprocessing beforehand
+    // we opt not to use crossover in the LLVMFuzzerMutate and instead have a second crossover pass,
+    // though it is likely an error for fuzzers to provide custom mutators but not custom crossovers
+    let custom_mutator = unsafe {
+        LLVMCustomMutator::mutate_unchecked(StdMOptMutator::new(
+            &mut state,
+            havoc_mutations_no_crossover(), // TODO .merge(tokens_mutations()),
+            7,
+            5,
+        )?)
+    };
+    let std_mutator_no_mutate = StdScheduledMutator::with_max_stack_pow(havoc_crossover(), 3);
+
+    let cm_power = StdPowerMutationalStage::new(custom_mutator, &edges_observer);
+    let cm_power = SkippableStage::new(cm_power, |_| mutator_status.custom_mutation.into());
+    let cm_std_power = StdMutationalStage::new(std_mutator_no_mutate);
+    let cm_std_power = SkippableStage::new(cm_std_power, |_| mutator_status.std_no_mutate.into());
+
+    // a custom crossover is defined
+    // while the scenario that a custom crossover is defined without a custom mutator is unlikely
+    // we handle it here explicitly anyways
+    let custom_crossover = unsafe {
+        LLVMCustomMutator::crossover_unchecked(StdScheduledMutator::with_max_stack_pow(
+            havoc_mutations_no_crossover(), // TODO .merge(tokens_mutations()),
+            3,
+        ))
+    };
+    let std_mutator_no_crossover = StdMOptMutator::new(
+        &mut state,
+        havoc_mutations_no_crossover(), // TODO .merge(tokens_mutations()),
+        7,
+        5,
+    )?;
+
+    let cc_power = StdMutationalStage::new(custom_crossover);
+    let cc_power = SkippableStage::new(cc_power, |_| mutator_status.custom_crossover.into());
+    let cc_std_power = StdPowerMutationalStage::new(std_mutator_no_crossover, &edges_observer);
+    let cc_std_power =
+        SkippableStage::new(cc_std_power, |_| mutator_status.std_no_crossover.into());
+
+    // unfortunately, we cannot support grimoire and also support custom mutators
```

Comment:
Ah I see WRT grimoire.
I think it should be doable somehow, no?

Reply:
I'm unsure. From what I recall, the .bytes() and .target_bytes() methods were weird for Grimoire in the past. I'll see if I can't finagle something.